opendatahub-io/odh-model-controller
# Add monitoring reconcilliation to odh-controller.

Dependent on:  
* https://github.com/red-hat-data-services/odh-manifests/pull/279
* https://github.com/red-hat-data-services/odh-deployer/pull/291

This change introduces a new reconciler to manage monitoring rbac for monitoring ServingRuntimes.

Test suite is also fixed up, and unsued/commented code chunks are removed.

## How Has This Been Tested?
Locally tested, and via live build

## Testing Instructions: 
Live build: quay.io/hukhan/rhods-operator-live-catalog:1.20.8-5439

Steps to test:
* Login to an OCP cluster via `oc` as cluster admin
* `git clone https://gitlab.cee.redhat.com/data-hub/olminstall.git && cd olminstall`
* `./setup.sh -t operator -i quay.io/hukhan/rhods-operator-live-catalog:1.20.8-54399`
* Wait for the modelmesh monitoring stack pods to come up
* Confirm the prometheus operator comes up succesfully, and prometheus instance is deployed 
  * `oc get pods redhat-ods-monitoring` 

Log in to the Prometheus instance by using the route obtained from: 
```bash
$ ROUTE=$(oc get routes -n redhat-ods-monitoring rhods-model-monitoring --template={{.spec.host}})
```

You should see ONLY 2 targets: 

![image](https://user-images.githubusercontent.com/10904967/206805008-ab332933-4b80-4158-b60d-a2e9d92e8138.png)

Now navigate to the dashboard as any user, create a Data Science Project called `my_project`. 

Confirm no monitoring rolebinding exists: 

```bash
$ oc get rolebinding  prometheus-ns-access -n my_project
Error from server (NotFound): rolebindings.rbac.authorization.k8s.io "prometheus-ns-access" not found
```

Setup Storage, then create a model server: 
* ` git clone https://github.com/opendatahub-io/modelmesh-serving.git && cd modelmesh-serving/quickstart`
* `oc apply -f sample-minio.yaml -n my_project 
* Deploy Model Server via the UI
* Navigate to $(ROUTE) via browser, confirm no targets exist (give it ~1min), we don't expect a target to show up since ther
* Deploy a model via Dashboard UI 

Confirm rolebinding is created: 

```bash
~ $ oc get rolebinding  prometheus-ns-access -n my-project
NAME                   ROLE                               AGE
prometheus-ns-access   ClusterRole/prometheus-ns-access   25m
~ $ 
```

* Navigate to $(ROUTE) via browser, confirm a new target is added: 
![image](https://user-images.githubusercontent.com/10904967/206806109-cf476d24-ec00-4ca0-9612-f9d4554b59af.png)

Give it a min or so, once model is created, MM will scale up the modelmesh runtime pod. 

* Confirm there are no errors in the prometheus pods in  `oc get pods redhat-ods-monitoring` 
  * look for permission errors in particular
  * prometheus pod may report permission denied logs for the datascience project, this is expected, and should stop once a model server is created, this is because modelmesh-controller creates service monitors before serving runtimes are created, a future iteration can account for this by reconciling on service monitors, not service runtime instead. 

We also expect to see rolebindings in `openshift-monitoring` (for federation) and in `redhat-ods-applications` (MM creates a service monitor in the controller namespace, so we add rbac here as well, though it doesn't do anything yet)
```bash 
~ $ oc get rolebindings -A | grep prometheus-ns-access
```

* Delete your model/model server via ui, we expect the rolebinding to be cleaned up when there are no serving runtimes in the namespace 
* try updating the rolebinding, or deleting it, we expect the controller to recreate it, or revert any user changes 
* Once targets are up, ensure you can run queries to ha proxy or model mesh metrics endpoints 

## Merge criteria:
